### PR TITLE
Preserve prefixes for unloaded items and unloaded prefixes

### PR DESCRIPTION
**Pull request by Tyfyter**
The benefits of this addition are listed in #3250, the details of its implementation are roughly as described in [this comment](https://github.com/tModLoader/tModLoader/issues/3250#issuecomment-1428491828_) by @Chicken-Bones

### CODE REVIEW

**Comment by Tyfyter** [comment]
Concerns that a displayed name would cause "item names [to] become pretty silly" were not taken into account as the expected course of action is to remove the prefix, either by re-enabling the unloaded mod or reforging the item, and other unloaded content looks similarly out of place.

File: patches/tModLoader/Terraria/ModLoader/Default/UnloadedGlobalItem.cs [hunk 1]
```diff
@@ -9,6 +9,11 @@ public class UnloadedGlobalItem : GlobalItem
 	[CloneByReference] // safe to share between clones, because it cannot be changed after creation/load
 	internal IList<TagCompound> data = new List<TagCompound>();
 
+	[CloneByReference] // safe to share between clones, because it cannot be changed after creation/load
+	internal string modPrefixMod = null;
+	[CloneByReference] // safe to share between clones, because it cannot be changed after creation/load
+	internal string modPrefixName = null;
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/Default/UnloadedGlobalItem.cs]: No need for comment. `string` is immutable.

File: patches/tModLoader/Terraria/ModLoader/Default/UnloadedPrefix.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+public sealed class UnloadedPrefix : ModPrefix
+{
+	public override LocalizedText DisplayName => Language.GetText("Mods.ModLoader.Unloaded");
+	public override IEnumerable<TooltipLine> GetTooltipLines(Item item)
+	{
+		UnloadedGlobalItem unloadedGlobalItem = item.GetGlobalItem<UnloadedGlobalItem>();
+		yield return new TooltipLine("UnloadedPrefix", $"{unloadedGlobalItem.modPrefixMod}/{unloadedGlobalItem.modPrefixName}") {
+			IsModifier = true,
+			OverrideColor = new(215, 123, 186)
+		};
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/Default/UnloadedPrefix.cs]: Please add screenshots of some unloaded prefixed item tooltips to the PR description.
Also include one of an unloaded item with an unloaded prefix

**Review by Chicken-Bones — CHANGES_REQUESTED** [review]

File: patches/tModLoader/Terraria/ModLoader/Default/UnloadedPrefix.cs (hunk 2 above)
Comment (Tyfyter) [reply patches/tModLoader/Terraria/ModLoader/Default/UnloadedPrefix.cs]: At this time, if unloaded items retain prefixes they still do not support actually having them — which is necessary for them to be displayed — an image of a loaded item with an unloaded prefix has been attached.

**Comment by JavidPack** [comment]
Code was added to preserve prefixes on UnloadedItem. Previously even vanilla prefixes would be lost on unloaded items.

**Review by Chicken-Bones — APPROVED** [review]

File: patches/tModLoader/Terraria/ModLoader/Default/UnloadedGlobalItem.cs [hunk 3]
```diff
@@ -9,6 +9,11 @@ public class UnloadedGlobalItem : GlobalItem
 	[CloneByReference] // safe to share between clones, because it cannot be changed after creation/load
 	internal IList<TagCompound> data = new List<TagCompound>();
 
+	[CloneByReference]
+	internal string modPrefixMod = null;
+	[CloneByReference]
+	internal string modPrefixName = null;
```
Comment (direwolf420) [inline patches/tModLoader/Terraria/ModLoader/Default/UnloadedGlobalItem.cs]:
```suggestion
	[field: CloneByReference]
	public string ModPrefixMod { get; internal set; } = null;
	[field: CloneByReference]
	public string ModPrefixName { get; internal set; } = null;
```
To follow the same design as UnloadedItem.ModName/ItemName. There are use cases for this to be accessible by other mods (including one of my mods). Obviously the case change should be applied to all places where it's used.

Comment (JavidPack) [reply patches/tModLoader/Terraria/ModLoader/Default/UnloadedGlobalItem.cs]: Done.

**Pull request merged**